### PR TITLE
[masters] Add extractor

### PR DESCRIPTION
**Pull request by m4tu4g**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [ x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [ x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [ x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [ x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

New extractor addition, [Masters](https://www.masters.com), closes https://github.com/yt-dlp/yt-dlp/issues/3240

### CODE REVIEW

File: yt_dlp/extractor/masters.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+from __future__ import unicode_literals
+from .common import InfoExtractor
+from ..utils import url_or_none
+
+
+class MastersIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?masters\.com/en_US/watch/[\d]{4}-[\d]{2}-[\d]{2}/(?P<id>[0-9]+)/(.*)\.html'
+    _TESTS = [{
+        'url': 'https://www.masters.com/en_US/watch/2022-04-07/16493755593805191/sungjae_im_thursday_interview_2022.html',
```
Comment (pukkandan) [inline yt_dlp/extractor/masters.py:9]: is the date in URL the upload date? If so, extract it

File: yt_dlp/extractor/masters.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
+from __future__ import unicode_literals
+from .common import InfoExtractor
+from ..utils import url_or_none
+
+
+class MastersIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?masters\.com/en_US/watch/[\d]{4}-[\d]{2}-[\d]{2}/(?P<id>[0-9]+)/(.*)\.html'
```
Comment (pukkandan) [inline yt_dlp/extractor/masters.py:7]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?masters\.com/en_US/watch/)?P<date>\d{4}-\d{2}-\d{2})/(?P<id>\d+)'
```

File: yt_dlp/extractor/masters.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            'title': 'Sungjae Im: Thursday Interview 2022',
+            'thumbnail': r're:^https?://.*\.jpg$',
+        }
+    }]
+
+    _CONTENT_API_URL = "https://www.masters.com/relatedcontent/rest/v2/masters_v1/en/content/masters_v1_{video_id}_en"
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        content_resp = self._download_json(self._CONTENT_API_URL.format(video_id=video_id), video_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/masters.py:22]:
```suggestion
        content_resp = self._download_json(
            f'https://www.masters.com/relatedcontent/rest/v2/masters_v1/en/content/masters_v1_{video_id}_en',
            video_id)
```

File: yt_dlp/extractor/masters.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        formats = self._extract_m3u8_formats(content_resp['media']['m3u8'], video_id, 'mp4')
+        self._sort_formats(formats)
+
+        return {
+            'id': video_id,
+            'title': content_resp["title"],
```
Comment (pukkandan) [inline yt_dlp/extractor/masters.py:28]:
```suggestion
            'title': content_resp.get('title'),
```

File: yt_dlp/extractor/masters.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+            'formats': formats,
+            'thumbnail': url_or_none(content_resp['images'][0]['large']),
```
Comment (pukkandan) [inline yt_dlp/extractor/masters.py:30]:
```suggestion
            'thumbnail': traverse_obj(content_resp, ('images', 0, 'large'), expected_type=url_or_none),
```
if there are multiple thumbnails, extract them all

**Review by pukkandan — CHANGES_REQUESTED** [review]
> * [ x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md?rgh-link-date=2022-04-08T17%3A02%3A54Z#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md?rgh-link-date=2022-04-08T17%3A02%3A54Z#yt-dlp-coding-conventions)

It doesn't look like u did :(

File: yt_dlp/extractor/masters.py (hunk 1 above)
Comment (m4tu4g) [reply yt_dlp/extractor/masters.py:9]: Yes. will return that as well

**Comment by m4tu4g** [comment]
> > * [ x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md?rgh-link-date=2022-04-08T17%3A02%3A54Z#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md?rgh-link-date=2022-04-08T17%3A02%3A54Z#yt-dlp-coding-conventions)
> 
> It doesn't look like u did :(

hehe, first PR and all these blunders

**Comment by m4tu4g** [comment]
@pukkandan 

Changed conventions, PR to be reviewed

File: yt_dlp/extractor/masters.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+            'title': 'Sungjae Im: Thursday Interview 2022',
+            'upload_date': '20220407',
+            'thumbnail': r're:^https?://.*\.jpg$',
+        }
+    }]
+
+    _CONTENT_API_URL = "https://www.masters.com/relatedcontent/rest/v2/masters_v1/en/content/masters_v1_{video_id}_en"
+
+    def _real_extract(self, url):
+        video_id, upload_date = self._match_id(url), self._match_valid_url(url).group('date')
```
Comment (pukkandan) [inline yt_dlp/extractor/masters.py:25]:
```suggestion
        video_id, upload_date = self._match_valid_url(url).group('id', 'date')
```

File: yt_dlp/extractor/masters.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+        content_resp = self._download_json(
+            f'https://www.masters.com/relatedcontent/rest/v2/masters_v1/en/content/masters_v1_{video_id}_en',
+            video_id)
+        formats = self._extract_m3u8_formats(traverse_obj(content_resp, ('media', 'm3u8')), video_id, 'mp4')
```
Comment (pukkandan) [inline yt_dlp/extractor/masters.py:29]: _extract_m3u8_formats_and_subtitles

File: yt_dlp/extractor/masters.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        self._sort_formats(formats)
+
+        thumbnail_data = traverse_obj(content_resp, ('images', 0))
+        thumbnails = [{
+            'url': thumbnail_data.get(item)
+        } for item in thumbnail_data or []]
```
Comment (pukkandan) [inline yt_dlp/extractor/masters.py:35]:
```suggestion
        thumbnails = [{'id': name, 'url': url} for name, url in traverse_obj(content_resp, ('images', 0)) or []]
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/masters.py (hunk 7 above)
Comment (m4tu4g) [reply yt_dlp/extractor/masters.py:29]: There doesn't seem to be any subs info coming from response so left it like that on seeing other extractors. Should I return None?

Comment (pukkandan) [reply yt_dlp/extractor/masters.py:29]: it is recommended to always use _and_subs varinat of the functions for newer extractors
```suggestion
        formats, subtitles = self._extract_m3u8_formats_and_subtitles(
            traverse_obj(content_resp, ('media', 'm3u8')), video_id, 'mp4')
```
the return `subtitles` too. It's fine if it is not present

Comment (m4tu4g) [reply yt_dlp/extractor/masters.py]: Done but had to revert back to old convension for **thumbnails** (getting **too many values to unpack** error)

File: yt_dlp/extractor/masters.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    def _real_extract(self, url):
+        video_id, upload_date = self._match_valid_url(url).group('id', 'date')
+        content_resp = self._download_json(
+            f'https://www.masters.com/relatedcontent/rest/v2/masters_v1/en/content/masters_v1_{video_id}_en',
+            video_id)
+        formats, subtitles = self._extract_m3u8_formats_and_subtitles(traverse_obj(content_resp, ('media', 'm3u8')), video_id, 'mp4')
+        self._sort_formats(formats)
+
+        thumbnail_data = traverse_obj(content_resp, ('images', 0))
+        thumbnails = [{'id': name, 'url': thumbnail_data.get(name)} for name in thumbnail_data or []]
```
Comment (pukkandan) [inline yt_dlp/extractor/masters.py:33]:
```suggestion
        thumbnails = [{'id': name, 'url': url} for name, url in traverse_obj(content_resp, ('images', 0), default={}).items()]
```

File: yt_dlp/extractor/masters.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+        'info_dict': {
+            'id': '16493755593805191',
+            'ext': 'mp4',
+            'title': 'Sungjae Im: Thursday Interview 2022',
+            'upload_date': '20220407',
+            'thumbnail': r're:^https?://.*\.jpg$',
+        }
+    }]
+
+    _CONTENT_API_URL = "https://www.masters.com/relatedcontent/rest/v2/masters_v1/en/content/masters_v1_{video_id}_en"
```
Comment (pukkandan) [inline yt_dlp/extractor/masters.py:22]: unused
```suggestion
```

File: yt_dlp/extractor/masters.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
+from __future__ import unicode_literals
+from .common import InfoExtractor
+from ..utils import (
+    unified_strdate,
+    traverse_obj,
```
Comment (pukkandan) [inline yt_dlp/extractor/masters.py:5]:
```suggestion
    traverse_obj,
    unified_strdate,
```

File: yt_dlp/extractor/masters.py [hunk 12]
```diff
@@ -0,0 +1,39 @@
+from __future__ import unicode_literals
+from .common import InfoExtractor
+from ..utils import (
+    traverse_obj,
+    unified_strdate,
+)
+
+
+class MastersIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?masters\.com/en_US/watch/(?P<date>\d{4}-\d{2}-\d{2})/(?P<id>\d+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/masters.py:10]: is this always en_US?

Comment (m4tu4g) [reply yt_dlp/extractor/masters.py:10]: yes, couldn't find any other

**Pull request merged**